memif-new/Dockerfile

```diff
@@ -0,0 +1,15 @@
+FROM ghcr.io/edwarnicke/govpp/vpp:20.09 as go
+COPY --from=golang:1.15.3-buster /usr/local/go/ /go
+ENV PATH ${PATH}:/go/bin
+ENV GO111MODULE=on
+ENV CGO_ENABLED=0
+ENV GOBIN=/bin
+ENV GOPROXY=https://goproxy.io,direct
+RUN rm -r /etc/vpp
+RUN go get github.com/go-delve/delve/cmd/dlv@latest
+
+FROM go as build
+WORKDIR /workdir
+
+FROM ghcr.io/edwarnicke/govpp/vpp:20.09 as runtime
+
```

Comment:
OK... let's try to get this to a testable debuggable form.

We need to get to a place where with simple docker commands we can:

1.  Build
2. Run automated tests
3. Debug those tests

From the Dockerfile point of view you want to have multiple stages:

1. A 'go' stage with go installed
```
FROM ghcr.io/edwarnicke/govpp/vpp:${VPP_VERSION} as go
COPY --from=golang:1.16.3-buster /usr/local/go/ /go
ENV PATH ${PATH}:/go/bin
ENV GO111MODULE=on
ENV CGO_ENABLED=0
ENV GOBIN=/bin
RUN go get github.com/go-delve/delve/cmd/dlv@v1.6.0
```

2. A 'build' stage.
```
FROM go as build
WORKDIR /build
COPY go.mod go.sum ./
RUN go mod download
RUN go build ./...
```

3. Add a 'test' stage.
```
FROM build as test
CMD go test -test.v ./...
```

4. Add a 'debug' stage.
```
FROM test as debug
WORKDIR /build/internal/tests/
CMD dlv -l :40000 --headless=true --api-version=2 test -test.v .
```

So you can run the tests with:

```bash
docker run --rm $(docker build -q --target test .)
```

Or debug the tests with:

```bash
docker run --rm -p 40000:40000 $(docker build -q --target debug .)
```

For this to be useful, you will need to run self contained tests to try things.  Given this, it becomes possible to reasonably debug at fairly fine grained detail what is happening.

Reply:
```bash
docker exec -it $(docker ps -q) /bin/bash 
```

Reply:
https://gobyexample.com/testing

Reply:
https://github.com/edwarnicke/govpp/blob/a045363b0c3639c14685dc550b9592ad6ef96310/binapi/interface/interface_rpc.ba.go#L25

Reply:
https://github.com/networkservicemesh/sdk-vpp/blob/66ee1708a54a353383f61f752240979e610ff114/pkg/networkservice/up/common.go#L50-L55

Reply:
https://github.com/networkservicemesh/cmd-forwarder-vpp/blob/cd71461fa8a8442fd6dbee251af8cbc453abba83/internal/tests/suite_memif_test.go#L152-L161

Reply:
And for latter when you want to check things: https://github.com/stretchr/testify#assert-package